### PR TITLE
Remove argparse utils

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

The argparse utilities, namely `Trainer.add_argparse_args` and `Trainer.parse_argparse_args`, were introduced early on in Lighting at a time when the Trainer had a relatively simple argument structure. Over time, the variety of types and combinations thereof has grown, leading to a more complex and fragile parsing logic of default values and types, e.g., 

https://github.com/Lightning-AI/lightning/blob/5340d960b9427dcfe5fb953cac0e86763d423546/src/lightning/pytorch/utilities/argparse.py#L143-L155

https://github.com/Lightning-AI/lightning/blob/5340d960b9427dcfe5fb953cac0e86763d423546/src/lightning/pytorch/utilities/argparse.py#L240-L265

We are removing this logic now in a breaking change for Lightning 2.0. The reasons being:
- Lots of hacky, fragile logic, hard to maintain
- There are better 3rd party solutions out there: jsonargparse/click/hydra/fire …
- For the simple scripts, a hand-written argparse does the job as well
- For several releases, we had the the argparse utils already marked as "legacy" in our docs
- We practically get no questions or feedback about this feature, indicating that its popularity has decreased
- It has always been cumbersome to both capture and parse the arguments of both Trainer and LightningModule into one single parser. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @justusschock

### CODE REVIEW

File: docs/source-pytorch/common_usecases.rst [hunk 1]
```diff
@@ -32,13 +32,6 @@ Customize and extend Lightning for things like custom hardware or distributed st
    :button_link: model/build_model.html
    :height: 100
 
-.. displayitem::
-   :header: Configure hyperparameters from the CLI
-   :description: Enable basic CLI with Lightning.
-   :col_css: col-md-12
-   :button_link: common/hyperparameters.html
-   :height: 100
-
```
Comment (awaelchli) [inline docs/source-pytorch/common_usecases.rst:41]: I was thinking, maybe I should keep this and only explain the bare use of argparse with Trainer, LM, etc.

File: docs/source-pytorch/common/trainer.rst [hunk 2]
```diff
@@ -118,38 +118,15 @@ So you can run it like so:
...
-
-.. note::
-    If you want to stop a training run early, you can press "Ctrl + C" on your keyboard.
-    The trainer will catch the ``KeyboardInterrupt`` and attempt a graceful shutdown, including
-    running accelerator callback ``on_train_end`` to clean up memory. The trainer object will also set
-    an attribute ``interrupted`` to ``True`` in such cases. If you have a callback which shuts down compute
-    resources, for example, you can conditionally run the shutdown logic for only uninterrupted runs.
+If you want to stop a training run early, you can press "Ctrl + C" on your keyboard.
+The trainer will catch the ``KeyboardInterrupt`` and attempt a graceful shutdown, including
+running accelerator callback ``on_train_end`` to clean up memory. The trainer object will also set
```
Comment (carmocca) [inline docs/source-pytorch/common/trainer.rst:127]: We don't do this as it's described.

```suggestion
The trainer will catch the ``KeyboardInterrupt`` and attempt a graceful shutdown. The trainer object will also set
```

File: docs/source-pytorch/common/trainer.rst [hunk 3]
```diff
@@ -118,38 +118,15 @@ So you can run it like so:
...
+an attribute ``interrupted`` to ``True`` in such cases. If you have a callback which shuts down compute
+resources, for example, you can conditionally run the shutdown logic for only uninterrupted runs.
```
Comment (carmocca) [inline docs/source-pytorch/common/trainer.rst:129]:
```suggestion
an attribute ``interrupted`` to ``True`` in such cases. If you have a callback which shuts down compute
resources, for example, you can conditionally run the shutdown logic for only uninterrupted runs by overriding :meth:`lightning.pytorch.Callback.on_exception`.
```

File: docs/source-pytorch/common_usecases.rst (hunk 1 above)
Comment (carmocca) [reply docs/source-pytorch/common_usecases.rst:41]: It wouldn't hurt as an educational example, but I wouldn't go deep into it. I suggest doing that in a follow-up and also have some brief examples with popular argument parsing libraries.

File: examples/pl_domain_templates/generative_adversarial_net.py [hunk 4]
```diff
@@ -226,8 +226,8 @@ def main(args: Namespace) -> None:
     # ------------------------
     # If use distributed training  PyTorch recommends to use DistributedDataParallel.
     # See: https://pytorch.org/docs/stable/nn.html#torch.nn.DataParallel
-    dm = MNISTDataModule.from_argparse_args(args)
-    trainer = Trainer.from_argparse_args(args)
+    dm = MNISTDataModule()
+    trainer = Trainer(accelerator="gpu", devices=1)
```
Comment (carmocca) [inline examples/pl_domain_templates/generative_adversarial_net.py:230]: `examples/pl_domain_templates/semantic_segmentation.py` needs to be updated too

File: examples/pl_domain_templates/generative_adversarial_net.py [hunk 5]
```diff
@@ -239,15 +239,8 @@ def main(args: Namespace) -> None:
     cli_lightning_logo()
     parser = ArgumentParser()
 
-    # Add program level args, if any.
-    # ------------------------
-    # Add LightningDataLoader args
-    parser = MNISTDataModule.add_argparse_args(parser)
     # Add model specific args
     parser = GAN.add_argparse_args(parser)
```
Comment (carmocca) [inline examples/pl_domain_templates/generative_adversarial_net.py:243]: I suggest taking the code inside this method and putting it here for simplicity. The implementation can also remove the `use_argument_group=False` code path

File: examples/pl_domain_templates/reinforce_learn_ppo.py [hunk 6]
```diff
@@ -466,8 +465,6 @@ def main(args) -> None:
     seed_everything(0)
 
     parent_parser = argparse.ArgumentParser(add_help=False)
-    parent_parser = Trainer.add_argparse_args(parent_parser)
-
     parser = PPOLightning.add_model_specific_args(parent_parser)
```
Comment (carmocca) [inline examples/pl_domain_templates/reinforce_learn_ppo.py:468]: Same suggestion

File: src/lightning/pytorch/utilities/argparse.py [hunk 7]
```diff
@@ -15,101 +15,36 @@
...
 from contextlib import suppress
 from functools import wraps
-from typing import Any, Callable, cast, Dict, List, Tuple, Type, TypeVar, Union
+from typing import Any, Callable, cast, Type, TYPE_CHECKING, TypeVar, Union
 
-import lightning.pytorch as pl
-from lightning.pytorch.utilities.parsing import str_to_bool, str_to_bool_or_int, str_to_bool_or_str
-from lightning.pytorch.utilities.types import _ADD_ARGPARSE_RETURN
+if TYPE_CHECKING:
+    import lightning.pytorch as pl
```
Comment (carmocca) [inline src/lightning/pytorch/utilities/argparse.py:25]: This is unused

```suggestion
```

File: src/lightning/pytorch/utilities/argparse.py [hunk 8]
```diff
@@ -15,101 +15,36 @@
...
-                # which then becomes True here.
-
-                v = True
-
-        modified_args[k] = v
-    return Namespace(**modified_args)
-
-
-def parse_env_variables(cls: _ARGPARSE_CLS, template: str = "PL_%(cls_name)s_%(cls_argument)s") -> Namespace:
+def _parse_env_variables(cls: _ARGPARSE_CLS, template: str = "PL_%(cls_name)s_%(cls_argument)s") -> Namespace:
```
Comment (carmocca) [inline src/lightning/pytorch/utilities/argparse.py:31]: We don't need `_ARGPARSE_CLS` anymore

```suggestion


def _parse_env_variables(cls: Type, template: str = "PL_%(cls_name)s_%(cls_argument)s") -> Namespace:
```

File: src/lightning/pytorch/CHANGELOG.md [hunk 9]
```diff
@@ -267,6 +267,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Removed the unused `lightning.pytorch.utilities.metrics.metrics_to_scalars` function ([#16681](https://github.com/Lightning-AI/lightning/pull/16681))
 
+- Removed legacy argparse utilities ([#16708](https://github.com/Lightning-AI/lightning/pull/16708))
```
Comment (carmocca) [inline src/lightning/pytorch/CHANGELOG.md:270]: We can also remove the code in `src/lightning/pytorch/utilities/parsing.py`

File: src/lightning/pytorch/utilities/argparse.py (hunk 7 above)
Comment (awaelchli) [reply src/lightning/pytorch/utilities/argparse.py]: It was used for the definition of _ARGPARSE_CLS

File: src/lightning/pytorch/CHANGELOG.md (hunk 9 above)
Comment (awaelchli) [reply src/lightning/pytorch/CHANGELOG.md:270]: Done

**Review by Borda — APPROVED** [review]
